File: lib/to_mdx/parsers/open_document/draw/enhanced_geometry.rb
# frozen_string_literal: true

require_relative "../base"
require_relative "../mixins/has_formula"
require_relative "../mixins/positionable_element"
require_relative "../mixins/svg"

module ToMdx
  module Parsers
    module Draw
      class EnhancedGeometry < Base
        include Mixins::HasFormula
        include Mixins::PositionableElement
        include Mixins::Svg

        def initialize(element = nil, context = {})
          super(element, context)
          @draw_style_name = context[:styles].use(element&.attributes&.[]("draw:style-name"))
        end

        def empty?
          false
        end

        def to_mdx
          # TODO if it's just a rectangle that fills the entire parent text area / custom shape, just style the custom shape instead
          return "" if @element.attributes["draw:type"] == "ooxml-rect"
          wrap_with_svg_tag("<path d=\"#{svg_path}\" />", @draw_style_name)
        end

        def to_s
          # Enhanced geometry elements don't contribute text content
          ""
        end

        register_for "draw:enhanced-geometry"
      end
    end
  end
end

File: src/parsers/open-document/draw/enhanced_geometry.ts
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";
import { FormulaEvaluator } from "../mixins/formula_evaluator.ts";

declare module "../../base_element.ts" {
  interface ParseContext {
    viewBox?: [number, number, number, number];
  }
}

export class EnhancedGeometry extends BaseElement {
  private drawStyleName: Style;
  readonly drawType: string;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.drawStyleName = context.styles.use(this.attr("draw:style-name"));
    this.drawType = this.attr("draw:type") || "";
  }

  empty(): boolean {
    return false;
  }

  toMdx(): string {
    // TODO It seems like the paths don't contain, for example, arrowheads.
    //   We probably need to handle this by looking at specific draw types, and manually emitting the svg.
    //
    // switch (this.drawType) {
    //   case "mso-spt32":
    //     return `...`;
    // }

    const svgPath = this.hasFormula.svgPath();
    if (!svgPath) {
      return "";
    }

    const classAttr = this.drawStyleName.empty() ? "" : ` className="${this.drawStyleName}"`;
    const content = this.contentfulChildren()
      .map((child) => child.toMdx())
      .join("");
    return `<path${classAttr} d="${svgPath}" />${content}`;
  }

  toString(): string {
    return "";
  }

  private get hasFormula(): FormulaEvaluator {
    return new FormulaEvaluator(this);
  }

  static {
    BaseElement.registerFor.call(this, "draw:enhanced-geometry");
  }
}
